### PR TITLE
[java] New rule in migrating ruleset: ForLoopCanBeForeach

### PR DESCRIPTION
**Pull request by oowekyala**
The rule detects for loops over arrays, lists and iterators that can be safely replaced with a foreach loop. I compared it to IntelliJ's built in inspection to improve it, and it performs quite well! 

The code of the rule is quite scary though, sorry about that...

### CODE REVIEW

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    <test-code>
+        <description>Consider iterators</description>
+        <expected-problems>1</expected-problems>
+        <code><![CDATA[
+        class Foo {
+          void loop() {
+            for (Iterator<DataFlowNode> i = path.iterator(); i.hasNext();) {
+              DataFlowNode inode = i.next();
+              if (inode.getVariableAccess() == null) {
+                continue;
```
Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml]: if we had this same example, but where doing a `i.remove()` here, then the code wouldn't be possible to rewrite as a foreach loop, but the rule would flag it non the less

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/migrating/ForLoopCanBeForeachRule.java [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+        if (indexDecl == null) {
+            return super.visit(node, data);
+        }
+
+
+        List<NameOccurrence> occurrences = indexDecl.getValue();
+        VariableNameDeclaration index = indexDecl.getKey();
+
+        if ("Iterator".equals(index.getTypeImage()) && isReplaceableIteratorLoop(indexDecl, guardCondition, update)) {
+            addViolation(data, node);
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/migrating/ForLoopCanBeForeachRule.java]: When we add a violation, do we need to investigate further? Or could we simply return here?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/migrating/ForLoopCanBeForeachRule.java [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+        if (indexDecl == null) {
+            return super.visit(node, data);
+        }
+
+
+        List<NameOccurrence> occurrences = indexDecl.getValue();
+        VariableNameDeclaration index = indexDecl.getKey();
+
+        if ("Iterator".equals(index.getTypeImage()) && isReplaceableIteratorLoop(indexDecl, guardCondition, update)) {
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/migrating/ForLoopCanBeForeachRule.java]: There is a `net.sourceforge.pmd.lang.java.typeresolution.TypeHelper` utility class, you could use here:

    TypeHelper.isA(index, Iterator.class)

It uses typeresolution, if the type is available and as a fallback the type image.
The we should of course enable typeresolution for this rule.

File: pmd-java/src/main/resources/rulesets/java/migrating.xml [hunk 4]
```diff
@@ -477,4 +477,32 @@ public class MyTest {
...
   </rule>
 
-</ruleset>
\ No newline at end of file
+    <rule name="ForLoopCanBeForeach"
+          language="java"
+          since="6.0"
+          message="This 'for' loop can be replaced by a 'foreach' loop"
+          class="net.sourceforge.pmd.lang.java.rule.migrating.ForLoopCanBeForeachRule"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_migrating.html#forloopcanbeforeach">
```
Comment (adangel) [inline pmd-java/src/main/resources/rulesets/java/migrating.xml]: I've seen this now for some other rules: For completeness (it probably has no real usage, only if you happen to work on a code base < java 1.5), we should add `minimumLanguageVersion="1.5"`

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml (hunk 1 above)
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml]: You're right, I'll add a check

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/migrating/ForLoopCanBeForeachRule.java (hunk 2 above)
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/migrating/ForLoopCanBeForeachRule.java]: We can absolutely return here

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/migrating/ForLoopCanBeForeachRule.java (hunk 3 above)
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/migrating/ForLoopCanBeForeachRule.java]: Nice!

File: pmd-java/src/main/resources/rulesets/java/migrating.xml (hunk 4 above)
Comment (oowekyala) [inline pmd-java/src/main/resources/rulesets/java/migrating.xml]: Will do

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml (hunk 1 above)
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml]: I've pushed a commit which fixes that, though it only considers the "remove" method. Do you know of other methods that should be blacklisted?

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml [hunk 5]
```diff
@@ -225,4 +226,42 @@
...
+              }
+            }
+          }
+        }
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Do not report iterator loop if we can't find iterator decl</description>
```
Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml]: I'm not sure this is to be expected... sounds more as a limitation to the current implementation... am I missing something?

Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml]: Hmm yeah I think I intended it as "don't report if we can't find *iterator* decl" though the test code addresses a missing *iterable* déclaration 
I think there might be a few edge cases which would mess with the rule in that case. For instance, if the iterator is declared as a field in a parent class (symbol table wouldn't be able to resolve it) and `next ` was already called as a side effect eg of a constructor. Then the loop would not loop through all elements of the iterate.  Might be purely theoretical though

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml (hunk 1 above)
Comment (jsotuyod) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml]: Nope, the iterator only has those 3 methods: `remove`, `hasNext` and `next`,  and being able to use `remove` without a `ConcurrentModificationException` is pretty much the only valid reason to use an iterator instead of a foreach

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml (hunk 5 above)
Comment (jsotuyod) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml]: But the current typeresolution code *can* actually resolve the type of inherited fields (no need to get their declaration!)... so we can actually know if `path` is `Iterable` provided a complete classpath.

The only reason for not being able to detect it would be an incomplete classpath (missing classes), which would yield `null` as the type for path.

If we depended less on symbol table and more on type resolution directly, more of these scenarios would be addressed.

Comment (oowekyala) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml]: Hum well the expression `path.iterator()` is parsed as a primary prefix with image `path.iterator` and a primary suffix for the arguments. I think we cannot get the type of `path` there, only that of `path.iterator()`, right?

Comment (jsotuyod) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml]: You are right, we need #497 for this... it will have to wait then.

**Comment by adangel** [comment]
Merged! Thanks!
